File: RadarSDK/RadarRouteMatrix.m
//
//  RadarRouteMatrix.m
//  Library
//
//  Copyright © 2021 Radar Labs, Inc. All rights reserved.
//

#import "RadarRouteMatrix.h"
#import "RadarRoute+Internal.h"
#import "RadarRouteMatrix+Internal.h"

@implementation RadarRouteMatrix

- (nullable instancetype)initWithMatrix:(nullable NSArray<NSArray<RadarRoute *> *> *)matrix {
    self = [super init];
    if (self) {
        _matrix = matrix;
    }
    return self;
}

- (nullable instancetype)initWithObject:(_Nonnull id)object {
    if (![object isKindOfClass:[NSArray class]]) {
        return nil;
    }

    NSArray *rows = (NSArray *)object;

    NSMutableArray<NSMutableArray<RadarRoute *> *> *matrix = [NSMutableArray arrayWithCapacity:rows.count];

    for (int i = 0; i < rows.count; i++) {
        NSArray *col = rows[i];
        NSMutableArray<RadarRoute *> *routes = [NSMutableArray arrayWithCapacity:col.count];
        for (int j = 0; j < col.count; j++) {
            RadarRoute *route = [[RadarRoute alloc] initWithObject:col[j]];
            routes[j] = route;
        }
        matrix[i] = routes;
    }

    return [[RadarRouteMatrix alloc] initWithMatrix:matrix];
}

- (RadarRoute *_Nullable)routeBetweenOriginIndex:(NSUInteger)originIndex destinationIndex:(NSUInteger)destinationIndex {
    if (originIndex >= self.matrix.count) {
        return nil;
    }

    NSArray<RadarRoute *> *routes = self.matrix[originIndex];

    if (destinationIndex >= routes.count) {
        return nil;
    }

    return routes[destinationIndex];
}

- (NSArray *)arrayValue {
    NSMutableArray<NSMutableArray<NSDictionary *> *> *rows = [NSMutableArray arrayWithCapacity:self.matrix.count];
    for (int i = 0; i < self.matrix.count; i++) {
        NSArray<RadarRoute *> *routes = self.matrix[i];
        NSMutableArray<NSDictionary *> *col = [NSMutableArray arrayWithCapacity:routes.count];
        for (int j = 0; j < routes.count; j++) {
            RadarRoute *route = routes[j];
            col[j] = [route dictionaryValue];
        }
        rows[i] = col;
    }
    return rows;
}

@end

File: RadarSDK/RadarRouteMatrix.swift
//
//  RadarRouteMatrix.swift
//  RadarSDK
//
//  Copyright © 2021 Radar Labs, Inc. All rights reserved.
//

import Foundation

/// Represents routes between multiple origins and destinations.
///
/// Backs the `RadarRouteMatrix` interface declared in `RadarRouteMatrix.h` and
/// `RadarRouteMatrix+Internal.h`, so Objective-C callers are unaffected by the move to Swift.
///
/// - SeeAlso: https://radar.com/documentation/api#matrix
@objc(RadarRouteMatrix)
class RadarRouteMatrix: NSObject {

    /// Rows are origins and columns are destinations. Declared nonnull in
    /// `RadarRouteMatrix+Internal.h`, so a missing matrix is an empty one rather than nil.
    @objc public let matrix: [[RadarRouteObjc]]

    /// Keeps the hand-written Objective-C initializer working after the implementation moved to Swift.
    @objc(initWithMatrix:)
    init(matrix: [[RadarRouteObjc]]) {
        self.matrix = matrix
    }

    /// Parses the `matrix` field of a `/route/matrix` response.
    ///
    /// Rows keep their position so the indexes stay aligned with the request: a row that is not
    /// an array becomes an empty row, and an entry `RadarRoute` cannot parse is dropped. The
    /// Objective-C implementation instead raised an exception on either input, because it
    /// inserted the nil route straight into an `NSMutableArray`.
    @objc
    convenience init?(object: Any) {
        guard let rows = object as? [Any] else {
            return nil
        }

        self.init(
            matrix: rows.map { row in
                guard let col = row as? [Any] else {
                    return []
                }
                return col.compactMap { RadarRouteObjc(object: $0) }
            }
        )
    }

    /// Returns the route between the specified origin and destination, or nil if either index is
    /// out of range.
    @objc(routeBetweenOriginIndex:destinationIndex:)
    func routeBetween(originIndex: UInt, destinationIndex: UInt) -> RadarRouteObjc? {
        guard originIndex < UInt(matrix.count) else {
            return nil
        }

        let routes = matrix[Int(originIndex)]

        guard destinationIndex < UInt(routes.count) else {
            return nil
        }

        return routes[Int(destinationIndex)]
    }

    @objc
    func arrayValue() -> [[[String: Any]]] {
        matrix.map { routes in routes.map { $0.dictionaryValue() } }
    }
}
